Use distroless to build Conformance image

wilsonehusin opened the pull request:
Signed-off-by: Wilson Husin <whusin@vmware.com>

<!--  Thanks for sending a pull request!  Here are some tips for you:

1. If this is your first time, please read our contributor guidelines: https://git.k8s.io/community/contributors/guide/first-contribution.md#your-first-contribution and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. Please label this pull request according to what type of issue you are addressing, especially if this is a release targeted pull request. For reference on required PR/issue labels, read here:
https://git.k8s.io/community/contributors/devel/sig-release/release.md#issuepr-kind-label
3. Ensure you have added or ran the appropriate tests for your PR: https://git.k8s.io/community/contributors/devel/sig-testing/testing.md
4. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
5. If the PR is unfinished, see how to mark it: https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
-->

#### What type of PR is this?

<!--
Add one of the following kinds:
/kind bug
/kind cleanup
/kind documentation
/kind feature
/kind design

Optionally add one or more of the following kinds if applicable:
/kind api-change
/kind deprecation
/kind failing-test
/kind flake
/kind regression
-->

/kind cleanup

#### What this PR does / why we need it:

In order to avoid noisy / false positive CVE image scanning, Conformance image is now be built based on Distroless.

#### Which issue(s) this PR fixes:
<!--
*Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
_If PR is about `failing-tests or flakes`, please post the related issues/tests in a comment and do not use `Fixes`_*
-->
Fixes #96692

#### Special notes for your reviewer:

```
❯ sonobuoy status --show-all
         PLUGIN                 NODE     STATUS   RESULT
            e2e               global   complete   passed
```

This deprecates `/run_e2e.sh` -- while the existing parameters will still be honored, users running the test should not call `/run_e2e.sh` and should instead call `/kubeconformance` (previously named `/gorunner`).

I think I have made potentially controversial decisions and I'd like to point them out for discussion:

1. `/run_e2e.sh` still exists, just that now it force invokes `/kubeconformance`. If you have been using Sonobuoy to run conformance testing per [cncf/k8s-conformance instructions](https://github.com/cncf/k8s-conformance/blob/master/instructions.md), this change should not affect you since [gorunner has been the default executor since Sonobuoy 0.16](https://github.com/vmware-tanzu/sonobuoy/commit/0e16b5ea9e5a048f14398024634033fddaf8b423).
1. the name `gorunner` somewhat collides with [k/release `go-runner`](https://github.com/kubernetes/release/tree/master/images/build/go-runner) which are two different things -- ~do folks have opinions on what the new name of conformance `gorunner`? I'm leaning towards reusing `run_e2e` without `.sh` suffix.~ UPDATE: I have chosen to use `kubeconformance` as the binary name for now.
1. in Dockerfile, I'm pointing to a specific version of `gcr.io/k8s-staging-build-image/debian-base:v2.1.3` because I'm out of clues on how other images are finding the `debian-base`, if at all -- ideally I don't want to add more work for whoever doing bumps on `debian-base` images that they need to bump yet-another-file in this repository.
    - referring to `debian:latest` from DockerHub wouldn't be the best solution either because (I think) this is the last image in k/k which still pulls images from DockerHub.
1. in Dockerfile, I _think_ we should be using `ENTRYPOINT` instead of `CMD`, but the fact that it hasn't been done made me question my own thoughts, would love to hear what others think.

Having said all the above, I think these are current blockers for this PR:

- [x] print deprecation warning for invoking `run_e2e.sh`, ask to use the compiled Go-based runner instead.
- [x] rename `/gorunner` into something else.
- [x] use alias instead of hardcoded `debian-base` reference.


#### Does this PR introduce a user-facing change?
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".

For more information on release notes see: https://git.k8s.io/community/contributors/guide/release-notes.md
-->
```release-note
[ACTION REQUIRED] Conformance image is now built with Distroless. Users running Conformance testing should rely on container entrypoint instead of manual invocation to `/run_e2e.sh` or `/gorunner`, as they are now deprecated and will be removed in 1.25 release. Invoking `ginkgo` and `e2e.test` are still supported through overriding entrypoint (docker) or defining container `spec.command` (kubernetes)
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

<!--
This section can be blank if this pull request does not require a release note.

When adding links which point to resources within git repositories, like
KEPs or supporting documentation, please reference a specific commit and avoid
linking directly to the master branch. This ensures that links reference a
specific point in time, rather than a document that may change over time.

See here for guidance on getting permanent links to files: https://help.github.com/en/articles/getting-permanent-links-to-files

Please use the following format for linking documentation:
- [KEP]: <link>
- [Usage]: <link>
- [Other doc]: <link>
-->
```docs

```

I'm not sure which "roles" and "hats" should be notified but since y'all were involved in the initial discussion 3 months ago, I figured to start with you all 😄  
/cc @justaugustus @BenTheElder

Commit:
Use distroless to build Conformance image

In order to avoid noisy / false positive CVE image scanning, Conformance
image is now be built based on Distroless.

Signed-off-by: Wilson Husin <whusin@vmware.com>

wilsonehusin commented:
just to see if I have made the robots overlord upset...

/test all

wilsonehusin commented:
Found this in the error log for pull-kubernetes-conformance-image-test:
```
2021-02-18T05:56:40.773115391Z stderr F 2021/02/18 05:56:40 Saving results at /tmp/results
2021-02-18T05:56:40.77318965Z stderr F 2021/02/18 05:56:40 failed to create log file /tmp/results/e2e.log: open /tmp/results/e2e.log: permission denied
```

`/tmp` should exist, but I'm guessing `/tmp/results` not existing is causing this. Will look at it tomorrow.

Commit:
Add /tmp/results directory with nonroot ownership

Signed-off-by: Wilson Husin <whusin@vmware.com>

wilsonehusin commented:
/test pull-kubernetes-conformance-image-test

vladimirvivien commented on cluster/images/conformance/Dockerfile:
```diff
@@ -14,12 +14,21 @@
 COPY e2e.test /usr/local/bin/
 COPY kubectl /usr/local/bin/
-COPY run_e2e.sh /run_e2e.sh
+COPY gorunner /run_e2e.sh
```
@wilsonehusin is that COPY correct ?

wilsonehusin replied:
@vladimirvivien Yeah this is to preserve folks invoking `/run_e2e.sh` so that it won't immediately break their automation. I'm working on getting a deprecation message printed on that, (4) above might be relevant -- if we set `ENTRYPOINT` for this image then we don't need to worry about this abstraction anymore.

BenTheElder replied:
This is still a breaking change either way (no shell utils available, script can't be sourced, etc.). Though I'm not sure if any users depended on that.

BenTheElder replied:
we should consider dropping this copy in the future, since we're storing two copies of the binary ..

wilsonehusin commented:
so I'm starting to wonder if the save results ever worked because the `run_e2e.sh` would fail silently if going into `/tmp/results` failed: 
https://github.com/kubernetes/kubernetes/blob/df8dee80c838bfb50532d1509c1a77a9c4463a11/cluster/images/conformance/run_e2e.sh#L33

wilsonehusin committed:
Add binaries to PATH, use ENTRYPOINT

Signed-off-by: Wilson E. Husin <wilsonehusin@gmail.com>

wilsonehusin commented:
/test pull-kubernetes-conformance-image-test

wilsonehusin commented:
> Found this in the error log for pull-kubernetes-conformance-image-test:
> 
> ```
> 2021-02-18T05:56:40.773115391Z stderr F 2021/02/18 05:56:40 Saving results at /tmp/results
> 2021-02-18T05:56:40.77318965Z stderr F 2021/02/18 05:56:40 failed to create log file /tmp/results/e2e.log: open /tmp/results/e2e.log: permission denied
> ```
> 
> `/tmp` should exist, but I'm guessing `/tmp/results` not existing is causing this. Will look at it tomorrow.

I reproduced this locally, seem to have gone away with `ENTRYPOINT` but it's still happening on Prow 🤔

wilsonehusin commented:
/test pull-kubernetes-conformance-image-test

wilsonehusin committed:
Use distroless/base:latest for Conformance image

Going with the guess that nonroot is the problem of creating
/tmp/results/e2e.log in container

Removing manual creation of /tmp/results as well

wilsonehusin commented:
/test pull-kubernetes-conformance-image-test

wilsonehusin commented:
```
Last State:     Terminated
  Reason:       ContainerStatusUnknown
  Message:      The container could not be located when the pod was deleted.  The container used to be Running 
```

🤔 is this a flake?

/test pull-kubernetes-conformance-image-test

wilsonehusin commented:
😲 it worked!!

/test all

wilsonehusin commented:
flake? 🤔 

```
State:          Terminated
  Reason:       Error
  Exit Code:    2
...
unable to retrieve container logs for containerd://6e8d752b18ec5e06ef71687a3acead9dbca520f71ef261fa6324c86bbd974c7b/workspace/_artifacts/logs 
```

/test pull-kubernetes-conformance-image-test

wilsonehusin commented:
yeah still flake it seems -- gonna proceed with the checklist for now.

wilsonehusin commented:
🙃 looks like this is the problem https://github.com/kubernetes/test-infra/blob/4dedc0e61a5540cace0b43c2d689112a75e6420c/experiment/kind-conformance-image-e2e.sh#L105-L110

wilsonehusin committed:
Use kubeconformance binary, deprecate gorunner and run_e2e

Actual name change will take place later as it involves changes to test-infra and other repositories

wilsonehusin commented:
alright I think this is _reasonably_ good state to be reviewed -- sorry for the earlier noise! let me know your thoughts @BenTheElder @justaugustus @vladimirvivien

BenTheElder commented on cluster/images/conformance/Makefile:
```diff
@@ -31,7 +31,7 @@ E2E_GO_RUNNER_BIN?=$(shell test -f $(LOCAL_OUTPUT_PATH)/go-runner && echo $(LOCA
 CLUSTER_DIR?=$(shell pwd)/../../../cluster/
 
-BASEIMAGE=debian:stable-slim
+BASEIMAGE=gcr.io/distroless/base:latest
```
I guess we don't have this in k8s.gcr.io?
Xref: https://github.com/kubernetes/kubernetes/issues/99238

wilsonehusin replied:
@BenTheElder sorry, re-reading this -- did you mean to imply that we should add this to k8s.gcr.io?

BenTheElder replied:
something to consider in a follow-up (or using go-runner?)

justaugustus replied:
+1 to using go-runner here instead!

wilsonehusin replied:
my initial attempt using go-runner didn't work, but i'll try again in a follow up!

BenTheElder reviewed (commented):
This seems like a good idea to me overall. It's still a potentially breaking change by design since additional binaries are not present.

wilsonehusin commented:
FYI @smarterclayton, I remember we had conversations about running Conformance in downstream contexts. Want to let you know in case you have thoughts 😄

wilsonehusin committed:
Use configurable BASEIMAGE and RUNNERIMAGE

Following example on /cluster/images/etcd/Makefile

wilsonehusin committed:
Add default value to KUBE_BASE_IMAGE_REGISTRY

wilsonehusin committed:
Use Distroless base instead of static

wilsonehusin commented:
/retest

wilsonehusin commented:
I'd like to check in again on this -- I know there is an unaddressed concern around breaking change (e.g. if users are running `source run_e2e.sh`) and I'm not sure what the path forward is with that. If they previously execute `./run_e2e.sh`, that should still work.

dims commented:
/assign @BenTheElder @spiffxp 

i'd personally lean on moving ahead with this!

BenTheElder commented on cluster/images/conformance/go-runner/main.go:
```diff
@@ -23,11 +23,15 @@ import (
 
 func main() {
+	if strings.Contains(os.Args[0], "run_e2e.sh") || strings.Contains(os.Args[0], "gorunner") {
+		log.Print("warn: calling test with e2e.test is deprecated, please rely on container manifest to invoke executable")
```
maybe we should set some rough deadline for when it will no longer be supported and tell the user?
thenw e can also stop storing it with this name.

wilsonehusin replied:
i thought about that but didn't feel like the right person to make such call 😅 do you think 1.25 is reasonable?

wilsonehusin replied:
okay I submitted 1.25 for now, let me know if we should change that 😄

BenTheElder reviewed (commented):
agree we should lean to move forward. left some comments.
thanks for working on this.

wilsonehusin commented:
/priority backlog
/triage accepted

wilsonehusin commented:
realized I put the wrong thing

/priority important-soon

wilsonehusin committed:
Add deadline for deprecation

dims commented:
looks like all the comments from @BenTheElder have been addressed

/approve
/lgtm

justaugustus commented:
/lgtm
/retest

wilsonehusin commented:
> TestExecPluginRotationViaInformer

seems unrelated

/retest

justaugustus commented:
Great work, Wilson! 💕🎉

wilsonehusin commented:
omg omg omg woooooooooo 🎉🎉
